### PR TITLE
Feat/talk2scholar hydra migration

### PR DESCRIPTION
**Pull request by ansh-info**
# For authors

#### Migration from config.py to Hydra Configuration System with Enhanced Test Coverage

1. **Summary of Changes**:
   - Migrated from static config.py and .env files to Hydra configuration system
   - Implemented YAML-based configurations for all components
   - Restructured configuration management in agents and tools
   - Added configuration support for frontend settings
   - Achieved 100% test coverage across all components

2. **Context and Motivation**:
   - Improved configuration management and maintainability
   - Enhanced flexibility for different environments
   - Better separation of concerns between configuration and code
   - More robust configuration validation
   - Easier configuration updates without code changes

## Implementation Details
1. Created new configuration structure:
```
configs/
├── agents/
│   └── talk2scholars/
│       ├── main_agent/
│       │   └── default.yaml
│       └── s2_agent/
│           └── default.yaml
├── app/
│   └── frontend/
│       └── default.yaml
└── tools/
    ├── search/
    ├── single_paper_recommendation/
    └── multi_paper_recommendation/
```

2. Component Updates:
   - Modified s2_agent.py for Hydra integration
   - Updated main_agent.py with new configuration system
   - Added tool-specific configurations
   - Implemented frontend configuration support

Test coverage includes:
1. Unit Tests:
   - Configuration loading and validation
   - Main Agent supervisor routing
   - S2 Tools functionality (search, recommendations, display)
   - State management
   - Configuration structure

2. Integration Tests:
   - End-to-end search workflow
   - Hydra configuration integration
   - Tool interactions

## Changed Files
1. Agents:
   - s2_agent.py
   - main_agent.py
2. Tools:
   - search.py
   - single_paper_rec.py
   - multi_paper_rec.py
   - display_results.py
3. Configuration:
   - Added multiple YAML files under configs/
   - Removed config.py
   - Updated .env handling
4. Tests:
   - Updated test fixtures and mocks
   - Added configuration testing
   - Enhanced test coverage to 100%

## Fixes # (issue) Mention the issue number.

Issue #61 

## Type of change
Please delete options that are not relevant.
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Please describe the tests you conducted to verify your changes. These may involve creating new test scripts or updating existing ones.
- [x] Added new test(s) in the ```tests``` folder
- [x] Added new function(s) to an existing test(s) (e.g.: ```tests/testX.py```)
- [ ] No new tests added (Please explain the rationale in this case)

## Checklist
- [x] My code follows the style guidelines mentioned in the Code/DevOps guides
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (e.g. MkDocs)
- [x] My changes generate no new warnings
- [x] I have added or updated tests (in the ```tests``` folder) that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# For reviewers

## Checklist pre-approval
- [ ] Is there enough documentation?  
- [ ] If a new feature has been added, or a bug fixed, has a test been added to confirm good behavior?
- [ ] Does the test(s) successfully test edge/corner cases?
- [ ] Does the PR pass the tests? (if the repository has continuous integration)

## Checklist post-approval
- [ ] Does this PR merge ```develop``` into ```main```? If so, please make sure to add a prefix (feat/fix/chore) and/or a suffix BREAKING CHANGE (if it's a major release) to your commit message.
- [ ] Does this PR close an issue? If so, please make sure to descriptively close this issue when the PR is merged.

## Checklist post-merge
- [ ] When you approve of the PR, merge and close it (Read this [article](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/about-merge-methods-on-github) to know about different merge methods on GitHub)
- [ ] Did this PR merge ```develop``` into ```main``` and is it suppose to run an automated release workflow (if applicable)? If so, please make sure to check under the "Actions" tab to see if the workflow has been initiated, and return later to verify that it has completed successfully.

### CODE REVIEW

File: aiagents4pharma/talk2scholars/agents/main_agent.py [hunk 1]
```diff
@@ -6,23 +6,21 @@
...
-from ..config.config import config
 from ..state.state_talk2scholars import Talk2Scholars
 
 logging.basicConfig(level=logging.INFO)
 logger = logging.getLogger(__name__)
 
-load_dotenv()
 
-def make_supervisor_node(llm: BaseChatModel) -> str:
+def make_supervisor_node(llm: BaseChatModel, cfg) -> str:
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/agents/main_agent.py:23]: Can you add the docstrings and a type hint for `cfg`?

File: aiagents4pharma/talk2scholars/agents/main_agent.py [hunk 2]
```diff
@@ -46,7 +46,7 @@ def supervisor_node(state: Talk2Scholars) -> Command[Literal["s2_agent", "__end_
         """
         logger.info("Supervisor node called")
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/agents/main_agent.py:47]: Would it be possible to add some additional information here that could be used to understand what the function was called with? By default, you would get the line number and method name so the INFO should provide more information than that.

File: aiagents4pharma/talk2scholars/agents/main_agent.py [hunk 3]
```diff
@@ -91,6 +92,15 @@ def get_app(thread_id: str, llm_model ='gpt-4o-mini') -> StateGraph:
     Returns:
         The compiled langraph app.
     """
+
+    # Load hydra configuration
+    logger.log(logging.INFO, "Load Hydra configuration for Talk2Scholars main agent.")
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/agents/main_agent.py:97]: For example, it might be useful to add in some of the hydra config values to the logging message.

File: aiagents4pharma/talk2scholars/configs/agents/talk2scholars/main_agent/default.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  "3. If unclear, ask for clarification\n"
+  "4. For multi-step tasks, focus on the immediate next step\n\n"
+  "Remember:\n"
+  "- Be decisive in your tool selection\n"
+  "- Focus on the immediate task\n"
+  "- Default to semantic_scholar_agent for any paper-finding tasks\n"
+  "- Ask for clarification if the request is ambiguous\n\n"
+  "When presenting paper search results, always use this exact format:\n\n"
+  "Remember to:\n"
+  "- Always remember to add the url\n"
```
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/configs/agents/talk2scholars/main_agent/default.yaml:37]:
```suggestion
  "- To always add the url\n"
```

File: aiagents4pharma/talk2scholars/configs/app/frontend/default.yaml
Comment (dmccloskey) [inline aiagents4pharma/talk2scholars/configs/app/frontend/default.yaml:1]: I think this is a good idea storing the front-end layout configs here. Is there a reason they are not used?

**Review by dmccloskey — CHANGES_REQUESTED** [review]
The migration to Hydra for handling configs looks very good 👍. Now that we are using hydra for confis, I would recommend adding additional information to the logging messages. In particular, ensuring that the log message captures more meaningful information that can be used to better understand what configs/arguments the function was called.

File: aiagents4pharma/talk2scholars/agents/main_agent.py (hunk 3 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/agents/main_agent.py:97]: Thank you for the detailed feedback! I've made the following updates based on your suggestions:

1. **Added `cfg` type hint and docstrings**  
   - The `make_supervisor_node` function now includes `cfg: Any` as a type hint.
   - Added a docstring explaining `cfg` as the configuration object.

File: aiagents4pharma/talk2scholars/agents/main_agent.py (hunk 2 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/agents/main_agent.py:47]: Thank you for the detailed feedback! I've made the following updates based on your suggestions:

1. **Added `cfg` type hint and docstrings**  
   - The `make_supervisor_node` function now includes `cfg: Any` as a type hint.
   - Added a docstring explaining `cfg` as the configuration object.

File: aiagents4pharma/talk2scholars/configs/agents/talk2scholars/main_agent/default.yaml (hunk 4 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/configs/agents/talk2scholars/main_agent/default.yaml]: I have made this change

File: aiagents4pharma/talk2scholars/configs/app/frontend/default.yaml
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/configs/app/frontend/default.yaml:1]: I was planning to cover this in the subsequent pr where I will also address this issue #76

Comment (ansh-info) [reply aiagents4pharma/talk2scholars/configs/app/frontend/default.yaml:1]: I have also created an Issue #97 To cover this

File: aiagents4pharma/talk2scholars/agents/main_agent.py [hunk 5]
```diff
@@ -44,9 +43,9 @@ def supervisor_node(state: Talk2Scholars) -> Command[Literal["s2_agent", "__end_
         Returns:
             Command[Literal["s2_agent", "__end__"]]: The command to execute next.
         """
-        logger.info("Supervisor node called")
+        logger.info("Supervisor node called with state: %s", state)
```
Comment (gurdeep330) [inline aiagents4pharma/talk2scholars/agents/main_agent.py:46]: This will end up printing the entire state on the screen, correct? So the log messages will keep getting longer with time.

File: pyproject.toml [hunk 6]
```diff
@@ -67,10 +67,12 @@ include = ["aiagents4pharma*"]
...
 aiagents4pharma = [
-            "configs/*",
-            "configs/talk2biomodels/agents/t2b_agent/*",
-            "configs/talk2biomodels/tools/ask_question/*",
-            "configs/talk2biomodels/tools/get_annotation/*"]
+  "configs/*",
+  "configs/talk2biomodels/agents/t2b_agent/*",
+  "configs/talk2biomodels/tools/ask_question/*",
+  "configs/talk2biomodels/tools/get_annotation/*",
+]
```
Comment (gurdeep330) [inline pyproject.toml:74]: I believe you will also need to add the path to the configs of T2S here. Do you mind double-checking? Thx!

**Review by gurdeep330 — CHANGES_REQUESTED** [review]

File: aiagents4pharma/talk2scholars/agents/main_agent.py (hunk 5 above)
Comment (ansh-info) [reply aiagents4pharma/talk2scholars/agents/main_agent.py:46]: Thanks for pointing that out! You were absolutely right—logging the full state would have made the logs unnecessarily large over time. I've updated the log message to include only the message count and current agent, which provides meaningful debugging information while keeping logs concise. Let me know if this looks good!

File: pyproject.toml (hunk 6 above)
Comment (ansh-info) [reply pyproject.toml:82]: Thank you, I have added the path to hydra config for talk2scholars

**Review by gurdeep330 — APPROVED** [review]
Thanks for addressing the reviews!

**Review by dmccloskey — APPROVED** [review]
nice work 💪
